Sources/ClientRuntime/Networking/Streaming/BufferedStream.swift

```diff
@@ -5,66 +5,142 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
-import Foundation
+import struct Foundation.Data
+import class Foundation.NSRecursiveLock
 
 /// A `Stream` implementation that buffers data in memory.
 /// The buffer size depends on the amount of data written and read.
 /// Note: This class is thread-safe.
 /// Note: if data is not read from the stream, the buffer will grow indefinitely until the stream is closed.
 ///       or reach the maximum size of a `Data` object.
 public class BufferedStream: Stream {
```

Comment:
This class is substantially changed internally, but it functions the same from the perspective of the caller.

Instead of polling intermittently to see if there is data to be read, this class now stores a continuation queue for callers wanting to read from the stream & resumes the continuations when it is appropriate to resume them.